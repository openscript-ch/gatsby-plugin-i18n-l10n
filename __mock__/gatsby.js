const React = require('react');
const gatsby = jest.requireActual('gatsby');
module.exports = {
  ...gatsby,
  graphql: jest.fn(),
  // these props are invalid for an `a` tag
  Link: jest.fn(({ activeClassName, activeStyle, getProps, innerRef, partiallyActive, ref, replace, to, ...rest }) =>
    React.createElement('a', {
      ...rest,
      href: to,
    }),
  ),
  StaticQuery: jest.fn(),
  useStaticQuery: jest.fn(),
};
