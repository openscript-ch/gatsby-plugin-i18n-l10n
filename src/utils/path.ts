import { posix as nodePath } from 'path';
import { PluginOptions } from 'gatsby';

export const handleTrailingSlash = (path: string, trailingSlashOption: PluginOptions['trailingSlash'] = 'always') => {
  if (path === '/' || path.length < 1) {
    return path;
  }

  const hasTrailingSlash = path.endsWith('/');

  switch (trailingSlashOption) {
    case 'always':
      return hasTrailingSlash ? path : `${path}/`;
    case 'never':
      return hasTrailingSlash ? path.slice(0, -1) : path;
    default:
      return path;
  }
};

export const trimRightSlash = (path: string) => {
  return path === '/' ? path : path.replace(/\/$/, '');
};

export const trimSlashes = (path: string) => {
  return path === '/' ? path : path.replace(/^\/|\/$/g, '');
};

export const replaceSegmentsWithSlugs = (path: string, slugs: Record<string, string>) => {
  const keys = Object.keys(slugs);
  if (keys.length > 0) {
    const exp = new RegExp(keys.join('|'), 'g');
    return path.replace(exp, (match) => slugs[match]);
  }
  return path;
};

export const addLocalePrefix = (path: string, locale: string, prefix: string, defaultLocale: string) => {
  return locale !== defaultLocale ? trimRightSlash(`/${nodePath.join(prefix, path)}`) : path;
};

export const translatePagePath = (path: string, slugs: Record<string, string>, locale: string, prefix: string, defaultLocale: string) => {
  let translatedPath = path;

  translatedPath = replaceSegmentsWithSlugs(translatedPath, slugs);
  translatedPath = addLocalePrefix(translatedPath, locale, prefix, defaultLocale);

  return translatedPath;
};

export const translatePagePaths = (path: string, options: PluginOptions) => {
  return options.locales.map((locale) => {
    const trimmedPath = trimRightSlash(path);
    const newPath = locale.slugs[trimmedPath] ?? trimmedPath;

    return { locale: locale.locale, path: addLocalePrefix(newPath, locale.locale, locale.prefix, options.defaultLocale) };
  });
};

export const parsePathPrefix = (path: string, defaultPrefix: string) => {
  if (path === '/') {
    return defaultPrefix;
  }

  // Regex literals are evaluated when the script is loaded,
  // whereas the RegExp instantiation is done when it's reached
  // during execution. Safari doesn't support look behinds,
  // which causes an error when the script is loaded. This is
  // only needed in SSG.
  // eslint-disable-next-line prefer-regex-literals
  const splitPathExpression = new RegExp('(?<=^/)\\w{2}(-\\w{2})?(?=/)', 'g');
  const splittedPath = path.match(splitPathExpression);
  return splittedPath && splittedPath[0] ? splittedPath[0] : defaultPrefix;
};

export const generatePageContextByPath = (path: string, options: PluginOptions) => {
  const defaultPrefix = options.locales.find((l) => l.locale === options.defaultLocale)?.prefix;
  const parsedPrefix = parsePathPrefix(path, defaultPrefix ?? '');
  const locale = options.locales.find((l) => l.prefix === parsedPrefix);

  return { locale: locale?.locale, prefix: locale?.prefix };
};
