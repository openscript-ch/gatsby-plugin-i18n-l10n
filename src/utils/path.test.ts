import { PluginOptions } from 'gatsby';
import { addLocalePrefix, generatePageContextByPath, handleTrailingSlash, parsePathPrefix, translatePagePaths, trimSlashes } from './path';

const options: PluginOptions = {
  defaultLocale: `en-US`,
  siteUrl: '',
  locales: [
    {
      locale: `en-US`,
      prefix: `en`,
      slugs: {},
      messages: {},
    },
    {
      locale: `de-CH`,
      prefix: `de`,
      slugs: {
        '/imprint/': '/impressum/',
      },
      messages: {},
    },
    {
      locale: `zh-CN`,
      prefix: `zh`,
      slugs: {},
      messages: {},
    },
  ],
  plugins: [],
};

describe('handleTrailingSlash', () => {
  it('should leave trailing slahes, when the option is set to always', () => {
    const expectation = handleTrailingSlash('https://example.com/', 'always');
    expect(expectation).toBe('https://example.com/');
  });

  it('should add trailing slahes, when the option is set to always', () => {
    const expectation = handleTrailingSlash('https://example.com', 'always');
    expect(expectation).toBe('https://example.com/');
  });

  it('should remove trailing slahes, when the option is set to never', () => {
    const expectation = handleTrailingSlash('https://example.com/', 'never');
    expect(expectation).toBe('https://example.com');
  });

  it('should not add trailing slahes, when the option is set to never', () => {
    const expectation = handleTrailingSlash('https://example.com', 'never');
    expect(expectation).toBe('https://example.com');
  });

  it('should return root paths', () => {
    const expectation = handleTrailingSlash('/', 'never');
    expect(expectation).toBe('/');
  });

  it('should return empty paths', () => {
    const expectation = handleTrailingSlash('', 'never');
    expect(expectation).toBe('');
  });

  it('should return empty paths', () => {
    const expectation = handleTrailingSlash('', 'never');
    expect(expectation).toBe('');
  });

  it('should return paths, when the option is set to ignore', () => {
    let expectation = handleTrailingSlash('https://example.com', 'ignore');
    expect(expectation).toBe('https://example.com');
    expectation = handleTrailingSlash('https://example.com/', 'ignore');
    expect(expectation).toBe('https://example.com/');
  });

  it('should default to always', () => {
    const expectation = handleTrailingSlash('https://example.com');
    expect(expectation).toBe('https://example.com/');
  });

  it('should never add trailing slashes to error pages', () => {
    let expectation = handleTrailingSlash('404.html', 'always');
    expect(expectation).toBe('404.html');
    expectation = handleTrailingSlash('https://example.com/403.html/403.html', 'always');
    expect(expectation).toBe('https://example.com/403.html/403.html');
  });
});

describe('trimSlashes', () => {
  it('should trim the slashes', () => {
    const trimmed = trimSlashes('/de/page/');
    expect(trimmed).toBe('de/page');
  });

  it("shouldn't trim single slash", () => {
    const trimmed = trimSlashes('/');
    expect(trimmed).toBe('/');
  });

  it('should trim only leading slash if there is no trailing slash', () => {
    const trimmed = trimSlashes('/de/page');
    expect(trimmed).toBe('de/page');
  });

  it('should trim only trailing slash if there is no leading slash', () => {
    const trimmed = trimSlashes('de/page/');
    expect(trimmed).toBe('de/page');
  });
});

describe('addLocalePrefix', () => {
  it("adds locale as prefix if it's not the default locale", () => {
    const prefixed = addLocalePrefix('/path/to/rome', 'it', 'it', 'en');
    expect(prefixed).toBe('/it/path/to/rome');
  });
  it("adds locale as prefix if it's not the default locale to index routes", () => {
    const prefixed = addLocalePrefix('/', 'it', 'it', 'en');
    expect(prefixed).toBe('/it/');
  });
  it("adds no locale if it's default locale", () => {
    const prefixed = addLocalePrefix('/path/to/rome', 'it', 'it', 'it');
    expect(prefixed).toBe('/path/to/rome');
  });
});

describe('parsePathPrefix', () => {
  it('returns the default prefix if there is no path', () => {
    const prefix = parsePathPrefix('/', 'de');
    expect(prefix).toBe('de');
  });
  it('returns the prefix from the path', () => {
    const prefix = parsePathPrefix('/en/pages/another', 'de');
    expect(prefix).toBe('en');
  });
  it('returns the default prefix on empty path', () => {
    const prefix = parsePathPrefix('./', 'de');
    expect(prefix).toBe('de');
  });
});

describe('translatePagePaths', () => {
  it('returns an array of translated page paths', () => {
    const path = '/imprint';

    const pagePathTranslations = translatePagePaths(path, options);
    expect(pagePathTranslations).toEqual([
      { locale: 'en-US', path: '/imprint/' },
      { locale: 'de-CH', path: '/de/impressum/' },
      { locale: 'zh-CN', path: '/zh/imprint/' },
    ]);
  });

  describe('generatePageContextByPath', () => {
    const path = '/imprint';

    it('should generate the page context by path', () => {
      const expectation = generatePageContextByPath(path, options);
      expect(expectation).toMatchSnapshot();
    });

    it('should default to empty prefix', () => {
      const customOptions = { ...options, defaultLocale: 'fr-FR' };
      const expectation = generatePageContextByPath(path, customOptions);
      expect(expectation).toMatchSnapshot();
    });
  });
});
