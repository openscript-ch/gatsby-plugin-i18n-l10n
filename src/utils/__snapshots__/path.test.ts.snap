// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translatePagePaths generatePageContextByPath should default to empty prefix 1`] = `
{
  "locale": undefined,
  "prefix": undefined,
}
`;

exports[`translatePagePaths generatePageContextByPath should generate the page context by path 1`] = `
{
  "locale": "en-US",
  "prefix": "en",
}
`;
